src/lines/MultiLine.sol

```diff
@@ -0,0 +1,190 @@
+    event LineMessageExpired(bytes32 indexed lineMsgId);
+    event LineMessageExecution(bytes32 indexed lineMsgId);
+
+    constructor(address dao, address registry, string memory name) BaseMessageLine(name) {
+        _transferOwnership(dao);
+        REGISTRY = ILineRegistry(registry);
+    }
+
+    function setURI(string calldata uri) external onlyOwner {
+        _setURI(uri);
+    }
+
+    function setToLine(uint256 _toChainId, address _toLineAddress) external onlyOwner {
+        _setToLine(_toChainId, _toLineAddress);
+    }
+
+    function setFromLine(uint256 _fromChainId, address _fromLineAddress) external onlyOwner {
+        _setFromLine(_fromChainId, _fromLineAddress);
+    }
+
+    function _toLine(uint256 toChainId) internal view returns (address l) {
+        l = toLineLookup[toChainId];
+        require(l != address(0), "!toLine");
+    }
+
+    function _fromLine(uint256 fromChainId) internal view returns (address) {
+        return fromLineLookup[fromChainId];
+    }
+
+    function _send(address, uint256 toChainId, address toDapp, bytes calldata message, bytes calldata params)
+        internal
+        override
+    {
+        RemoteCallArgs memory args = abi.decode(params, (RemoteCallArgs));
+        multiSend(
+            MultiSendArgs(
+                args.names,
+                toChainId,
+                toDapp,
+                message,
+                args.params,
+                args.fees,
+                args.salt,
+                args.expiration,
+                args.threshold
+            )
+        );
+    }
+
+    function multiSend(MultiSendArgs memory args) public payable {
+        require(args.toChainId != LOCAL_CHAINID(), "!toChainId");
+
+        address fromDapp = msg.sender;
+        LineMsg memory lineMsg = LineMsg({
+            fromChainId: LOCAL_CHAINID(),
+            toChainId: args.toChainId,
+            fromDapp: fromDapp,
+            toDapp: args.toDapp,
+            salt: args.salt,
+            message: args.message,
+            expiration: args.expiration,
+            threshold: args.threshold
+        });
+        bytes32 lineMsgId = hash(lineMsg);
+        bytes memory encoded = abi.encodeWithSelector(MultiLine.multiRecv.selector, lineMsg);
+
+        uint256 totalFee = 0;
+        for (uint256 i = 0; i < args.names.length; i++) {
+            string memory name = args.names[i];
+            uint256 fee = args.fees[i];
+            address line = REGISTRY.getLine(name);
+            require(line != address(0), "!name");
+            IMessageLine(line).send{value: fee}(args.toChainId, _toLine(args.toChainId), encoded, args.params[i]);
+            totalFee += fee;
+        }
+
+        require(totalFee == msg.value, "!fees");
+        emit LineMessageSent(lineMsgId, args.names, lineMsg);
+    }
+
+    function multiRecv(LineMsg calldata lineMsg) external payable {
+        address line = _msgLine();
+        require(REGISTRY.isTrustedLine(line), "!line");
```

Comment:
This line could be any one in REGISTRY, but not the lines in the arg sent on source chain.

Suggest to remove REGISTRY, and put line addresses and oracle in the contract level.

Reply:
addrs, threshold. set by msgdao.